subsys/net/l2/ethernet/ethernet.c

```diff
@@ -562,12 +573,14 @@ static int ethernet_send(struct net_if *iface, struct net_pkt *pkt)
 	ret = api->send(net_if_get_device(iface), pkt);
 	if (ret != 0) {
 		eth_stats_update_errors_tx(iface);
+		ethernet_remove_l2_header(pkt);
```

Comment:
Let me ask the same question than in the other PR.

I wonder if it is correct to just pull the L2 header, while the ethernet driver still holds a reference of the packet which has just been queued in the DMA engine, and maybe not yet sent. This is the case for the SAM E70 GMAC driver for example. Judging by the comments (https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/ethernet/eth_mcux.c#L90-L104) the MCUX driver might also do that on the future.

 I can't test now, but I'll test tonight.

Reply:
would be really nice if you could try.

Then either the driver would need to ref that buffer again, or we could do something like #11873 : l2 would know that driver will keep an internal queue. (It would only unlink it from pkt->frags, but would not touch the reference)

As it is the very first fragment, it's "frags" attribute will point to the rest of the list, which one is going to be pkt->frags after this removal. So maybe such driver need to store that very first buf pointer in addition to the pkt.

Reply:
If this issue is found in 1 specific driver then yes. If it's in 2+ drivers it's worth factorizing it.

Reply:
we almost need a separate set of fragments for the headers that doesn't form a complete packet from end to end.  The total packet is the aggregation of the headers plus data.  This would keep the all of this isolated.  The extra work would be having to iterate over a list of lists to get all of the data on transfer.

Reply:
> This would keep the all of this isolated.

Well, the whole idea of the refactor was to allow to reuse the same header buffer for all packets. That was a brave refactor by @tbursztyka, and well, sometimes things don't go the easy way it initially seemed. (No fingerpointing, we just should accept (I did) that if we don't adopt well-known solutions like lwIP, then we're effectively experimenting out, and experiments sometimes fail.)

Reply:
Reusing the same ll header buffer would work only on eth device that don't queue/dma anything. But it is possible. (thus the PR #11873) Though there are not completely solved issue around gptp. Anyway, that's an optimization for small systems, so it's not the priority atm.

I pushed a new version of the PR. But I bet @aurel32 you'll find issues about queuing so that will need to be addressed next.

Reply:
> Reusing the same ll header buffer would work only on eth device that don't queue/dma anything. But it is possible. (thus the PR #11873)

I wonder if we should really design the network stack with the option of supporting DMA or not. Most (all?) SoCs with Ethernet have DMA support. I agree that not all drivers have to use DMA, and that it's a good idea to keep a few very simple drivers for now to test and validate the network stack. Having DMA support in all Ethernet drivers should still be a long term goal for Zephyr. So I really think the network stack should be able to support DMA by default.

> I pushed a new version of the PR. But I bet @aurel32 you'll find issues about queuing so that will need to be addressed next.

I confirm that there are issues, actually just a few seconds after booting:

```
[00:00:00.000,069] <inf> i2c_sam_twihs: Device I2C_0 initialized
[00:00:00.001,249] <inf> eth_sam: MAC: fc:c2:3d:12:e9:21
[00:00:00.001,385] <inf> eth_sam: Queue 0 activated
[00:00:00.001,390] <inf> eth_sam_phy: Soft Reset of ETH PHY


uart:~$ [00:00:00.140,016] <inf> eth_sam_phy: PHYID: 0x221561 at addr: 0
[00:00:00.140,016] <inf> eth_sam_phy: PHYID: 0x221561 at addr: 0
uart:~$ ***** Booting Zephyr OS zephyr-v1.13.0-3341-gc18fe74b53 *****
[00:00:02.820,019] <inf> eth_sam_phy: common abilities: speed 100 Mb, full duplex
[00:00:02.820,221] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.820,240] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already (tx_completed():569)
[00:00:02.820,431] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already (tx_completed():569)
[00:00:02.820,449] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.820,566] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already by tx_completed():569 (tx_completed():569)
[00:00:02.820,583] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.820,700] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already by tx_completed():569 (tx_completed():569)
[00:00:02.820,719] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.821,431] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already (tx_completed():569)
[00:00:02.821,453] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
0:02.826,257] <inf> net_config: Initializing network
00:00:02.820,019] <inf> eth_sam_phy: common abilities: speed 100 Mb, full duplex
uart:~$ [00:00:02.826,281] <inf> net_config: IPv4 address: 10.243.120.200
[00:00:02.820,221] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.820,240] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already (tx_completed():569)
[00:00:02.820,431] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already (tx_completed():569)
[00:00:02.820,449] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.820,566] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already by tx_completed():569 (tx_completed():569)
[00:00:02.820,583] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.820,700] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already by tx_completed():569 (tx_completed():569)
[00:00:02.820,719] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.821,431] <err> net_pkt: *** ERROR *** frag 0x2040ca10 is freed already (tx_completed():569)
[00:00:02.821,453] <err> net_pkt: *** ERROR *** frag 0x2040ca28 is freed already (tx_completed():569)
[00:00:02.826,257] <inf> net_config: Initializing network
[00:00:02.826,281] <inf> net_config: IPv4 address: 10.243.120.200
uart:~$ Single-threaded dumb HTTP server waits for a connection on port 8080...

```

Reply:
The driver, when queuing, should reference the pkt once more, but also ref the first net_buf, and  keep it by its own means (out of pkt, since l2 is going to remove it from the list).